### PR TITLE
Add span links to dashboard UI

### LINKED ISSUE #2577: Aspire Dashboard: Traces: Span links aren't shown in span details
When dealing with messaging systems, you typically don't want spans trying to encapsulate the processing of messages that you don't have control over, so rather than having a parent/child relationship, links can be used instead.
The dashboard does not currently display links in the details view for spans. 
Also linked spans are not shown unless they have the same TraceID and it's unclear if that will always happen in a message based system.

We should :-
1. Add a section for linked spans to the details of a span- and make it transitive, so show the links to the currrent span, and from the current span (they may not be the same)
2. Consider adding linked spans to the trace view to show the other spans that current ones are linked to.

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/aspire/issues/2577

* Displays list of span links for a span.
* Dashboard attempts to link the link IDs to in-memory span. If it is present then a summary of the span is displayed, otherwise the span id is displayed.
* Displays attributes in a grid if present.
* View button takes you to the trace/span if available. If not available, then the "loading" dialog from metrics exemplar is shown.

~PR doesn't include showing backlinks, but they could be added in the future. Backlinks are trickier:~
Backlinks done.

~Ready for review but shouldn't be merged until metrics exemplar is merged and the base branch is changed to main. The PR reuses functionality that was added for metrics exemplars.~
Metrics exemplars merged.

**Span with two links, one with attributes:**

image

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/4805)

### CODE REVIEW

**Comment by mitchdenny** [comment]
@JamesNK I was playing around with this PR locally and I clicked on a link and got this:

image

Seems stuck unless I cancel. Other links work.

**Comment by mitchdenny** [comment]
> Did you get that in the stress playground app? I forced some span links that will never resolve to make sure the UI was working. Displaying the dialog and then never doing anything is correct.

That explains it then.

**Comment by JamesNK** [comment]
Now with backlinks 🚀🔥🎈

**Comment by JamesNK** [comment]
Review please 🙏

File: src/Shared/CircularBuffer.cs [hunk 1]
```diff
@@ -106,6 +113,10 @@ public void Insert(int index, T item)
                 }
                 data[0] = overflowItem;
             }
+            else if (internalIndex < _end && _end < _buffer.Count - 1)
+            {
+                data.Slice(internalIndex, _end - internalIndex).CopyTo(data.Slice(_end));
+            }
```
Comment (JamesNK) [inline src/Shared/CircularBuffer.cs:119]: Fixes an existing bug in CircularBuffer that new tests highlighted.

File: src/Aspire.Dashboard/Components/Controls/SpanDetails.razor [hunk 2]
```diff
@@ -109,6 +109,74 @@
...
+                <FluentDataGrid TGridItem="SpanLinkViewModel"
+                                Items="@FilteredSpanLinks"
+                                Style="width:100%"
+                                GenerateHeader="GenerateHeaderOption.Sticky"
+                                GridTemplateColumns="4fr 1fr">
+                    <TemplateColumn Title="@Loc[nameof(ControlsStrings.SpanDetailsSpanColumnHeader)]">
+                        @WriteSpanLink(context)
+                    </TemplateColumn>
+                    <TemplateColumn Title="@Loc[nameof(ControlsStrings.SpanDetailsDetailsColumnHeader)]">
+                        <FluentButton Appearance="Appearance.Lightweight" OnClick="@(() => OnViewDetailsAsync(context))">View</FluentButton>
```
Comment (adamint) [inline src/Aspire.Dashboard/Components/Controls/SpanDetails.razor:127]: Hardcoded string?

File: src/Aspire.Dashboard/Model/SpanLinkViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.Dashboard.Otlp.Model;
+
+namespace Aspire.Dashboard.Model;
+
+public sealed class SpanLinkViewModel
```
Comment (adamint) [inline src/Aspire.Dashboard/Model/SpanLinkViewModel.cs:8]: Could this be a record?

**Review by adamint — APPROVED** [review]

**Pull request merged**